crates/client/rpc/src/lib.rs

```diff
@@ -1058,17 +998,30 @@ where
         })?;
         let chain_id = Felt252Wrapper(self.chain_id()?.0);
 
-        let message = message.try_into().map_err(|e| {
-            error!("Failed to convert MsgFromL1 to UserTransaction: {e}");
-            StarknetRpcApiError::InternalServerError
-        })?;
+        let transaction = {
+            let calldata = std::iter::once(Felt252Wrapper::from(message.from_address).into())
+                .chain(message.payload.into_iter().map(|felt| Felt252Wrapper::from(felt).into()))
+                .collect();
+            let tx = starknet_api::transaction::L1HandlerTransaction {
+                version: TransactionVersion::ZERO,
+                nonce: Nonce(StarkFelt::ZERO),
+                contract_address: Felt252Wrapper::from(message.to_address).into(),
+                entry_point_selector: Felt252Wrapper::from(message.entry_point_selector).into(),
+                calldata: Calldata(Arc::new(calldata)),
+            };
+            let tx_hash = tx.compute_hash(chain_id, true);
+
+            // Hardcoded `paid_fee_on_l1` value as it is not relevant here
+            L1HandlerTransaction { tx, tx_hash, paid_fee_on_l1: Fee(10) }
+        };
 
-        let fee_estimate = self.do_estimate_message_fee(substrate_block_hash, message)?;
+        let fee_estimate = self.do_estimate_message_fee(substrate_block_hash, transaction)?;
 
         let estimate = FeeEstimate {
             gas_price: fee_estimate.0.try_into().map_err(|_| StarknetRpcApiError::InternalServerError)?,
-            gas_consumed: fee_estimate.2,
-            overall_fee: fee_estimate.1,
+            gas_consumed: FieldElement::from(fee_estimate.2),
+            overall_fee: FieldElement::from(fee_estimate.1),
+            unit: PriceUnit::Wei,
```

Comment:
I guess that will depend right ?

Reply:
Yup. Fees are still to be properly implemented.